File: golang/679.24-game.go

const EPS = 1e-6

func judgePoint24(cards []int) bool {
	fcards := make([]float64, len(cards))
	for i, card := range cards {
		fcards[i] = float64(card)
	}

	return dfsJudge(fcards)
}

func dfsJudge(cards []float64) bool {
	if len(cards) == 1 {
		return math.Abs(cards[0]-24.0) < EPS
	}

	for i := range len(cards) - 1 {
		for j := i + 1; j < len(cards); j++ {
			newCards := make([]float64, 0, 4)
			for k, c := range cards {
				if k != i && k != j {
					newCards = append(newCards, c)
				}
			}

			ops := make([]float64, 0, 4+2+1)
			ops = append(ops,
				cards[i]+cards[j],
				cards[i]*cards[j],
				cards[i]-cards[j],
				cards[j]-cards[i],
			)

			if cards[i] > EPS {
				ops = append(ops, cards[j]/cards[i])
			}

			if cards[j] > EPS {
				ops = append(ops, cards[i]/cards[j])
			}

			for _, op := range ops {
				if dfsJudge(append(newCards, op)) {
					return true
				}
			}
		}
	}

	return false
}

File: python/679.24-game.py
import math


class Solution:
    def judgePoint24(self, cards: list[int]) -> bool:
        EPS = 1e-7

        def dfs(cards: list[int | float]) -> bool:
            if len(cards) == 1:
                return math.isclose(cards[0], 24, rel_tol=EPS)

            for i in range(len(cards)):
                for j in range(i + 1, len(cards)):
                    a, b = cards[i], cards[j]
                    new_cards = cards[:i] + cards[i + 1 : j] + cards[j + 1 :]
                    ops = [a + b, a - b, b - a, a * b]
                    if b > EPS:
                        ops.append(a / b)
                    if a > EPS:
                        ops.append(b / a)

                    for op in ops:
                        new_cards.append(op)
                        if dfs(new_cards):
                            return True
                        new_cards.pop()

            return False

        return dfs(list(map(float, cards)))

